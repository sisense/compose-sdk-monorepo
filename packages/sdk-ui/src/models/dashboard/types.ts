import { CommonFiltersOptions } from '@/common-filters/types';
import { ColorPaletteTheme, TabberConfig } from '@/types';

export type {
  CommonFiltersOptions,
  CommonFiltersApplyMode,
  FiltersIgnoringRules,
} from '@/common-filters/types';

/**
 * Column layout of dashboard widgets
 */
export interface WidgetsPanelColumnLayout {
  columns: {
    widthPercentage: number;
    rows: {
      cells: {
        widthPercentage: number;
        /**
         * @privateRemarks
         * This value appears to be overwritten by the widget's height property.
         * Marking as optional and internal for now.
         * @internal
         */
        height?: number | string;
        widgetId: string;
      }[];
    }[];
  }[];
}

/**
 * Layout of dashboard widgets panel, which is a union of different layout algorithms
 */
export type WidgetsPanelLayout = WidgetsPanelColumnLayout;

/**
 * Widget ID
 */
export type WidgetId = string;
/**
 * Options for widgets in a dashboard
 *
 * For example, how common filters defined at the dashboard level should be applied to widgets.
 */
export type WidgetsOptions = Record<WidgetId, { filtersOptions?: CommonFiltersOptions }>;
/**
 * Options for TabberWidetss in a dashboard
 *
 * This property actually moves responsibility on the layout management from the tabber widgets to the dashboard,
 * storing all the tabbers configs in the single place
 * @internal
 */
export type TabbersOptions = Record<WidgetId, TabberConfig>;

/**
 * Style options for the dashboard.
 */
export type DashboardStyleOptions = {
  /** Collection of colors used to color various elements */
  palette?: ColorPaletteTheme;
  /** Background color */
  backgroundColor?: string;
  /** Width of the divider line between widgets */
  dividerLineWidth?: number;
  /** Divider line color */
  dividerLineColor?: string;
};
