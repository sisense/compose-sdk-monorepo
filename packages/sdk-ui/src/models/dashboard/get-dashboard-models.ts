import { HttpClient } from '@sisense/sdk-rest-client';
import { RestApi } from '../../api/rest-api.js';
import { translateDashboard } from './translate-dashboard.js';

export interface GetDashboardModelsOptions {
  /**
   * Dashboard title to search by
   *
   * Dashboard titles are not necessarily unique, so the result may contain multiple dashboards.
   */
  searchByTitle?: string;
  /**
   * {@inheritDoc GetDashboardModelOptions.includeWidgets}
   */
  includeWidgets?: boolean;
}

/** @internal */
export async function getDashboardModels(
  http: HttpClient,
  options: GetDashboardModelsOptions = {},
) {
  const { includeWidgets, searchByTitle } = options;
  const api = new RestApi(http);
  const expand: string[] = [];
  let fields = ['oid', 'title', 'datasource'];

  if (includeWidgets) {
    // Removes "fields" list due to it's conflict with "expand" parameter => API endpoint issue
    fields = [];
    expand.push('widgets');
  }

  const dashboards = await api.getDashboards({ fields, expand, searchByTitle });

  return dashboards.map((dashboard) => translateDashboard(dashboard));
}
