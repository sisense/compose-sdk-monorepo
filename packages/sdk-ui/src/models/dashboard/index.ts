export * from './types';
export * from './get-dashboard-model';
export * from './get-dashboard-models';
export * from './use-get-dashboard-model';
export * from './use-get-dashboard-models';
export * from './dashboard-model';
export * as dashboardModelTranslator from './dashboard-model-translator';
