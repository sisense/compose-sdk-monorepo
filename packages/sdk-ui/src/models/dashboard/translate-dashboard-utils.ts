import {
  type LayoutDto,
  type FilterDto,
  type CascadingFilterDto,
  isCascadingFilterDto,
} from '../../api/types/dashboard-dto';
import type { WidgetsPanelColumnLayout, WidgetsOptions } from './types';
import {
  CascadingFilter,
  Filter,
  createFilterFromJaql,
  FilterRelations,
  FilterRelationsModel,
  FilterRelationsModelNode,
} from '@sisense/sdk-data';
import { CommonFiltersApplyMode } from '@/common-filters/types';
import {
  combineFiltersAndRelations,
  convertFilterRelationsModelToRelationRules,
  isTrivialSingleNodeRelations,
} from '@/utils/filter-relations';
import { WidgetDto } from '@/widget-by-id/types';

export const translateLayout = (layout: LayoutDto): WidgetsPanelColumnLayout => ({
  columns: (layout.columns || []).map((c) => ({
    widthPercentage: c.width,
    rows: (c.cells || []).map((cell) => {
      const totalWidth = cell.subcells.reduce((acc, subcell) => acc + subcell.width, 0);
      return {
        cells: cell.subcells.map((subcell) => ({
          // If the total width of the subcells is less than 100, we increase width percentage to make the subcells fill the column
          widthPercentage: totalWidth < 100 ? subcell.width / (totalWidth / 100) : subcell.width,
          height: subcell.elements[0].height,
          widgetId: subcell.elements[0].widgetid,
        })),
      };
    }),
  })),
});

const createFilterFromFilterDto = (filterDto: FilterDto): Filter => {
  const filter: Filter = createFilterFromJaql(
    filterDto.jaql,
    filterDto.instanceid,
    filterDto.disabled,
    filterDto.locked,
  );
  return filter;
};

const createFilterFromCascadingFilterDto = (
  cascadingFilterDto: CascadingFilterDto,
): CascadingFilter => {
  const { levels, instanceid, disabled, locked } = cascadingFilterDto;

  const innerFilters = levels.map((level) => createFilterFromJaql(level, level.instanceid));
  return new CascadingFilter(innerFilters, { guid: instanceid, disabled, locked });
};

export function extractDashboardFilters(
  dashboardFilters: Array<FilterDto | CascadingFilterDto>,
  filterRelationsModel?: FilterRelationsModel | FilterRelationsModelNode,
): Filter[] | FilterRelations {
  const filters = dashboardFilters.map((f) =>
    isCascadingFilterDto(f) ? createFilterFromCascadingFilterDto(f) : createFilterFromFilterDto(f),
  );
  if (!filterRelationsModel) {
    return filters;
  }
  const filterRelations = convertFilterRelationsModelToRelationRules(filterRelationsModel, filters);
  if (!filterRelations || isTrivialSingleNodeRelations(filterRelations)) {
    return filters;
  }
  return combineFiltersAndRelations(filters, filterRelations);
}

export function translateWidgetsOptions(widgets: WidgetDto[] = []): WidgetsOptions {
  const widgetsOptionsMap: WidgetsOptions = {};

  widgets.forEach((widget: WidgetDto) => {
    widgetsOptionsMap[widget.oid] = {
      filtersOptions: {
        applyMode:
          widget.options?.dashboardFiltersMode === 'filter'
            ? CommonFiltersApplyMode.FILTER
            : CommonFiltersApplyMode.HIGHLIGHT,
        shouldAffectFilters: widget.options?.selector,
        ignoreFilters: {
          all: widget.metadata.ignore?.all,
          ids: widget.metadata.ignore?.ids,
        },
        forceApplyBackgroundFilters: true,
      },
    };
  });

  return widgetsOptionsMap;
}
