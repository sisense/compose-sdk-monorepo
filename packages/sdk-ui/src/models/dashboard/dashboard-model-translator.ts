import { DashboardProps } from '@/dashboard';
import { DashboardModel } from './dashboard-model.js';
import { widgetModelTranslator } from '@/models/widget/';
import { DashboardDto } from '@/api/types/dashboard-dto.js';
import { CompleteThemeSettings } from '@/types.js';
import { AppSettings } from '@/app/settings/settings.js';
import {
  extractDashboardFilters,
  translateLayout,
  translateTabbersOptions,
  translateWidgetsOptions,
} from './translate-dashboard-utils.js';
import { DataSource } from '@sisense/sdk-data';

/**
 * Translates {@link DashboardModel} to {@link DashboardProps}.
 *
 * @example
 * ```tsx
 * <Dashboard {...dashboardModelTranslator.toDashboardProps(dashboardModel)} />
 * ```
 * @group Fusion Assets
 * @fusionEmbed
 */
export function toDashboardProps(dashboardModel: DashboardModel): DashboardProps {
  const {
    title,
    dataSource,
    widgets: widgetModels,
    layoutOptions,
    filters,
    widgetsOptions,
    tabbersOptions,
    styleOptions,
  } = dashboardModel;
  return {
    title,
    defaultDataSource: dataSource,
    widgets: widgetModels.map(widgetModelTranslator.toWidgetProps),
    layoutOptions,
    config: {},
    filters,
    widgetsOptions,
    styleOptions,
    tabbersOptions,
  };
}

/**
 * Creates a new dashboard model from a dashboard DTO.
 *
 * @param dashboardDto - The dashboard DTO to be converted to a dashboard model
 * @param themeSettings - Optional theme settings
 * @param appSettings - Optional application settings
 * @internal
 */
export function fromDashboardDto(
  dashboardDto: DashboardDto,
  themeSettings?: CompleteThemeSettings,
  appSettings?: AppSettings,
): DashboardModel {
  const {
    oid,
    title,
    datasource,
    widgets: widgetDtoList,
    layout: layoutDto,
    filters: filterDtoList,
    filterRelations: filterRelationsDtoOptions,
    style,
  } = dashboardDto;

  const dataSource: DataSource = {
    title: datasource.title,
    type: datasource.live ? 'live' : 'elasticube',
  };
  const styleOptions = {
    ...(style?.palette ? { palette: { variantColors: style?.palette.colors } } : null),
  };

  const mergedThemeSettings = themeSettings
    ? {
        ...themeSettings,
        ...(styleOptions.palette ? { palette: styleOptions.palette } : null),
      }
    : themeSettings;
  const widgets =
    widgetDtoList?.map((widget) =>
      widgetModelTranslator.fromWidgetDto(widget, mergedThemeSettings, appSettings),
    ) || [];
  const widgetsPanelLayout = layoutDto ? translateLayout(layoutDto) : { columns: [] };
  const filterRelationsModel = filterRelationsDtoOptions?.[0]?.filterRelations;
  const filters = extractDashboardFilters(filterDtoList || [], filterRelationsModel);
  const widgetsOptions = translateWidgetsOptions(widgetDtoList);
  const tabbersOptions = translateTabbersOptions(widgetDtoList);

  const dashboardModel: DashboardModel = {
    oid,
    title,
    dataSource,
    styleOptions,
    widgets,
    layoutOptions: { widgetsPanel: widgetsPanelLayout },
    filters,
    widgetsOptions,
    tabbersOptions,
  };

  return dashboardModel;
}
