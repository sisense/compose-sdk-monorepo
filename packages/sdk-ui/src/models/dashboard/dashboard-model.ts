import { DataSource, Filter, FilterRelations } from '@sisense/sdk-data';
import { DashboardStyleOptions, TabbersOptions, WidgetModel } from '@/models';
import { type WidgetsOptions } from './types';
import { DashboardLayoutOptions } from '@/dashboard';

/**
 * Model of Sisense Fusion dashboard defined in the abstractions of Compose SDK.
 *
 * @group Fusion Assets
 * @fusionEmbed
 */
export interface DashboardModel {
  /**
   * Unique identifier of the dashboard.
   */
  oid: string;

  /**
   * Dashboard title.
   */
  title: string;

  /**
   * Dashboard data source details.
   */
  dataSource: DataSource;

  /**
   * Dashboard widget models.
   */
  widgets: WidgetModel[];

  /**
   * Dashboard layout options.
   *
   * @internal
   */
  layoutOptions: DashboardLayoutOptions;

  /**
   * Dashboard style options.
   *
   * @internal
   */
  styleOptions: DashboardStyleOptions;

  /**
   * Dashboard filters.
   *
   * @internal
   */
  filters: Filter[] | FilterRelations;

  /**
   * Dashboard options for each of the widgets.
   *
   * @internal
   */
  widgetsOptions: WidgetsOptions;

  /**
   * Tabber configurations for the dashboard.
   *
   * @internal
   */
  tabbersOptions?: TabbersOptions;
}
