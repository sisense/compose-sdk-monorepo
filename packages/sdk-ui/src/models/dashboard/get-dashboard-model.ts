import { HttpClient } from '@sisense/sdk-rest-client';
import { RestApi } from '../../api/rest-api';
import { DashboardModel, dashboardModelTranslator } from '@/models/dashboard';
import { CompleteThemeSettings } from '../../types';
import { AppSettings } from '@/app/settings/settings';
import { TranslatableError } from '@/translation/translatable-error';
import { getWidgetIdsFromDashboard } from '@/utils/extract-widget-ids';
import { WidgetDto } from '@/widget-by-id/types';

export interface GetDashboardModelOptions {
  /**
   * Boolean flag whether to include widgets in the dashboard model
   *
   * If not specified, the default value is `false`
   */
  includeWidgets?: boolean;

  /**
   * Boolean flag whether to include filters in the dashboard model
   *
   * If not specified, the default value is `false`
   */
  includeFilters?: boolean;
}

/**
 * Retrieves a dashboard model by its OID.
 *
 * @param http - The HTTP client
 * @param dashboardOid - The OID of the dashboard
 * @param options - The options to include widgets and filters in the dashboard model
 * @param themeSettings - Optional theme settings
 * @param appSettings - Optional application settings
 * @returns The dashboard model
 * @internal
 */
export async function getDashboardModel(
  http: HttpClient,
  dashboardOid: string,
  options: GetDashboardModelOptions = {},
  themeSettings?: CompleteThemeSettings,
  appSettings?: AppSettings,
): Promise<DashboardModel> {
  const { includeWidgets, includeFilters } = options;
  const api = new RestApi(http);
  const fields = ['oid', 'title', 'datasource', 'style'];

  const isWat = http.auth?.type === 'wat';

  if (includeWidgets) {
    fields.push('layout');
  }

  if (includeFilters) {
    fields.push('filters');
    fields.push('filterRelations');
  }

  const dashboard = await api.getDashboard(dashboardOid, { fields });

  if (!dashboard) {
    throw new TranslatableError('errors.dashboardWithOidNotFound', { dashboardOid });
  }

  if (includeWidgets) {
    let widgets: WidgetDto[];

    if (isWat) {
      // WAT (Web Access Token) authentication method restricts direct access to the
      // API endpoint `api/v1/dashboards/${dashboardOid}/widgets`.
      // As a workaround, widgets are fetched individually based on their references in the dashboard layout.
      const widgetIds = getWidgetIdsFromDashboard(dashboard);
      const fetchedWidgets = await Promise.all(
        widgetIds.map((id) => api.getWidget(id, dashboard.oid)),
      );
      widgets = fetchedWidgets.filter((widget): widget is WidgetDto => widget !== undefined);
    } else {
      // Fetch all widgets at once
      widgets = (await api.getDashboardWidgets(dashboardOid)) || [];
    }

    dashboard.widgets = widgets;
  }

  // Next could be replaced in future with expand 'style' in '/dashboards/' request
  // when lowest supported Sisense API version will support it.
  // Currently, expand 'style' work only from l2024.2.0 and cause crash in older.
  if (dashboard.style?.paletteId && !dashboard.style.palette) {
    const palettesDto = await api.getPalettes();
    const paletteDto = palettesDto?.find(({ _id }) => _id === dashboard.style?.paletteId);
    if (paletteDto) {
      dashboard.style.palette = {
        name: paletteDto.name,
        colors: paletteDto.colors,
      };
    }
  }

  return dashboardModelTranslator.fromDashboardDto(dashboard, themeSettings, appSettings);
}
