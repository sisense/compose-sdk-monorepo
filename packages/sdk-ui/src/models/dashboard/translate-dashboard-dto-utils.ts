import {
  BackgroundFilterJaql,
  BaseFilterJaql,
  CascadingFilter,
  DimensionalBaseMeasure,
  ExcludeMembersFilterJaql,
  Filter,
  FilterJaql,
  getColumnNameFromAttribute,
  getTableNameFromAttribute,
  IncludeMembersFilterJaql,
  isCascadingFilter,
  isMembersFilter,
  isRankingFilter,
  MembersFilter,
  RankingFilter,
  TurnOffMembersFilterJaql,
} from '@sisense/sdk-data';
import { CascadingFilterDto, FilterDto, LayoutDto } from '@/api/types/dashboard-dto';
import { ConditionFilterJaql } from '@sisense/sdk-data/dist/dimensional-model/filters/utils/types';
import { WidgetsPanelLayout } from '@/dashboard/types';

/**
 * Translates a {@link Filter} to a {@link FilterDto}.
 *
 * @param filter - The filter to translate.
 * @returns FilterDto
 * @internal
 */
export function filterToFilterDto(filter: Filter): FilterDto | CascadingFilterDto {
  return isCascadingFilter(filter)
    ? cascadingFilterToFilterDto(filter)
    : regularFilterToFilterDto(filter);
}

function regularFilterToFilterDto(filter: Filter): FilterDto {
  if (isMembersFilter(filter)) {
    return membersFilterToDto(filter);
  } else if (isRankingFilter(filter)) {
    return rankingFilterToDto(filter);
  }
  return getFilterBaseDto(filter);
}

function cascadingFilterToFilterDto(filter: CascadingFilter): CascadingFilterDto {
  return {
    isCascading: true,
    instanceid: filter.config.guid,
    disabled: filter.config.disabled,
    levels: filter.filters.map((f) => {
      const levelFilterDto = regularFilterToFilterDto(f);
      return {
        ...levelFilterDto.jaql,
        instanceid: levelFilterDto.instanceid,
      };
    }),
  };
}

function getFilterBaseDto(filter: Filter): FilterDto {
  const {
    config: { disabled },
  } = filter;
  const filterJaql: FilterJaql = Object.assign(Object.create(filter), {
    config: { ...filter.config, disabled: false },
  }).jaql(true);

  return {
    jaql: {
      ...filterJaql,
      table: getTableNameFromAttribute(filter.attribute),
      column: getColumnNameFromAttribute(filter.attribute),
    },
    disabled,
    instanceid: filter.config.guid,
    isCascading: false,
  };
}

function membersFilterToDto(filter: MembersFilter) {
  const filterDto = getFilterBaseDto(filter);

  (filterDto.jaql.filter as IncludeMembersFilterJaql).multiSelection =
    filter.config.enableMultiSelection;

  // Current implementation of MembersFilter missing original order of members
  if (filter.config.deactivatedMembers.length > 0) {
    if (filter.config.excludeMembers) {
      (filterDto.jaql.filter as ExcludeMembersFilterJaql).exclude.members = [
        ...filter.members,
        ...filter.config.deactivatedMembers,
      ];
    } else {
      (filterDto.jaql.filter as IncludeMembersFilterJaql).members = [
        ...filter.members,
        ...filter.config.deactivatedMembers,
      ];
    }
  }

  if (filter.config.backgroundFilter) {
    setInnerFilter(
      filterDto.jaql,
      regularFilterToFilterDto(filter.config.backgroundFilter).jaql.filter,
    );
  }

  if (filter.config.deactivatedMembers.length > 0) {
    setInnerFilter(filterDto.jaql, {
      exclude: { members: filter.config.deactivatedMembers },
      turnedOff: true,
    });
  }

  return filterDto;
}

function rankingFilterToDto(filter: RankingFilter): FilterDto {
  const filterDto = getFilterBaseDto(filter);

  // In case of formula we still have messing table and column names in context, but this not break filter
  if (filter.measure instanceof DimensionalBaseMeasure) {
    const resultFilter = filterDto.jaql.filter as ConditionFilterJaql;
    resultFilter.by!.table = getTableNameFromAttribute(filter.measure.attribute);
    resultFilter.by!.column = getColumnNameFromAttribute(filter.measure.attribute);

    resultFilter.rankingMessage = filter.measure.name;
  }

  return filterDto;
}

/**
 * Recursively sets the inner filter of a parent filter.
 *
 * @param parentFilter - The parent filter.
 * @param innerFilter - The inner filter.
 * @internal
 */
function setInnerFilter(
  parentFilter: FilterJaql,
  innerFilter: BaseFilterJaql | BackgroundFilterJaql | TurnOffMembersFilterJaql,
): void {
  if (parentFilter.filter) {
    return setInnerFilter(parentFilter.filter as FilterJaql, innerFilter);
  } else {
    parentFilter.filter = innerFilter;
  }
}

export function layoutToLayoutDto(layout: WidgetsPanelLayout): LayoutDto {
  const DEFAULT_HEIGHT = '500px';
  return {
    columns: layout.columns.map((column) => ({
      width: column.widthPercentage,
      cells: column.rows.map((row) => ({
        subcells: row.cells.map((cell) => ({
          width: cell.widthPercentage,
          elements: [
            {
              widgetid: cell.widgetId,
              height: cell.height ?? DEFAULT_HEIGHT,
            },
          ],
        })),
      })),
    })),
  };
}
