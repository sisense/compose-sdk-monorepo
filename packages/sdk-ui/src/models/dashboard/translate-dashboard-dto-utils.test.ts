import { CascadingFilter, filterFactory, measureFactory, MembersFilter } from '@sisense/sdk-data';
import * as DM from '../../__test-helpers__/sample-ecommerce';
import { filterToFilterDto } from '@/models/dashboard/translate-dashboard-dto-utils';
import { CascadingFilterDto, FilterDto } from '@/api/types/dashboard-dto';

describe('baseFilterToFilterDto', () => {
  it('should return correctly translated simple filter', () => {
    const filter = filterFactory.members(DM.Commerce.Gender, ['Male']);
    const result = filterToFilterDto(filter);
    expect(withoutInstanceId(result)).toMatchSnapshot();
  });

  it('should return correctly translated disabled filter', () => {
    const filter = filterFactory.members(DM.Commerce.Gender, ['Female'], {
      guid: 'filter-id',
      disabled: true,
    });
    const result = filterToFilterDto(filter);
    expect(withoutInstanceId(result)).toMatchSnapshot();
  });

  it('should return correctly translated members filter with disabled members', () => {
    const filter = filterFactory.members(DM.Commerce.Gender, ['Female']) as MembersFilter;
    filter.config.deactivatedMembers = ['Male'];
    const result = filterToFilterDto(filter) as FilterDto;

    expect(result.jaql.filter).toMatchSnapshot();
  });

  it('should return correctly translated exclude members filter', () => {
    const config = { excludeMembers: true };
    const filter = filterFactory.members(DM.Commerce.Gender, ['Female'], config);
    const result = filterToFilterDto(filter) as FilterDto;

    expect(result.jaql.filter).toMatchSnapshot();
  });

  it('should return correctly translated members filter with background filter', () => {
    const config = {
      excludeMembers: false,
      backgroundFilter: filterFactory.members(DM.Commerce.Gender, ['Female', 'Male']),
    };

    const filter = filterFactory.members(DM.Commerce.Gender, ['Female'], config);
    const result = filterToFilterDto(filter) as FilterDto;

    expect(result.jaql.filter.filter).toMatchSnapshot();
  });

  it('should return correctly translated ranking filter', () => {
    const filter = filterFactory.bottomRanking(
      DM.Commerce.AgeRange,
      measureFactory.sum(DM.Commerce.Revenue),
      5,
    );
    const result = filterToFilterDto(filter) as FilterDto;

    expect(result.jaql.filter).toMatchSnapshot();
  });

  it('should return correctly translated cascading filter', () => {
    const filter = new CascadingFilter(
      [
        filterFactory.members(DM.Commerce.BrandID, ['1'], { guid: 'BRAND_FILTER_ID' }),
        filterFactory.members(DM.Commerce.Gender, ['Female'], { guid: 'GENDER_FILTER_ID' }),
      ],
      { guid: 'CASCADING_FILTER_ID' },
    );
    const result = filterToFilterDto(filter);

    expect(result).toMatchSnapshot();
  });
});

/** Removes the instanceid from the filter dto */
function withoutInstanceId<T extends FilterDto | CascadingFilterDto>(
  filter: T,
): Omit<T, 'instanceid'> {
  // eslint-disable-next-line no-unused-vars
  const { instanceid, ...rest } = filter;
  return rest;
}
