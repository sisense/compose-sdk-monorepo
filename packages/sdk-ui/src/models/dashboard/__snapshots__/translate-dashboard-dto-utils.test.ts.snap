// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`baseFilterToFilterDto > should return correctly translated cascading filter 1`] = `
{
  "disabled": false,
  "instanceid": "CASCADING_FILTER_ID",
  "isCascading": true,
  "levels": [
    {
      "column": "Brand ID",
      "datatype": "numeric",
      "dim": "[Commerce.Brand ID]",
      "filter": {
        "members": [
          "1",
        ],
        "multiSelection": true,
      },
      "instanceid": "BRAND_FILTER_ID",
      "table": "Commerce",
      "title": "Brand ID",
    },
    {
      "column": "Gender",
      "datatype": "text",
      "dim": "[Commerce.Gender]",
      "filter": {
        "members": [
          "Female",
        ],
        "multiSelection": true,
      },
      "instanceid": "GENDER_FILTER_ID",
      "table": "Commerce",
      "title": "Gender",
    },
  ],
}
`;

exports[`baseFilterToFilterDto > should return correctly translated disabled filter 1`] = `
{
  "disabled": true,
  "isCascading": false,
  "jaql": {
    "column": "Gender",
    "datatype": "text",
    "dim": "[Commerce.Gender]",
    "filter": {
      "members": [
        "Female",
      ],
      "multiSelection": true,
    },
    "table": "Commerce",
    "title": "Gender",
  },
}
`;

exports[`baseFilterToFilterDto > should return correctly translated exclude members filter 1`] = `
{
  "exclude": {
    "members": [
      "Female",
    ],
  },
  "multiSelection": true,
}
`;

exports[`baseFilterToFilterDto > should return correctly translated members filter with background filter 1`] = `
{
  "members": [
    "Female",
    "Male",
  ],
  "multiSelection": true,
}
`;

exports[`baseFilterToFilterDto > should return correctly translated members filter with disabled members 1`] = `
{
  "filter": {
    "exclude": {
      "members": [
        "Male",
      ],
    },
    "turnedOff": true,
  },
  "members": [
    "Female",
    "Male",
  ],
  "multiSelection": true,
}
`;

exports[`baseFilterToFilterDto > should return correctly translated ranking filter 1`] = `
{
  "bottom": 5,
  "by": {
    "agg": "sum",
    "column": "Revenue",
    "datatype": "numeric",
    "dim": "[Commerce.Revenue]",
    "table": "Commerce",
    "title": "sum Revenue",
  },
  "rankingMessage": "sum Revenue",
}
`;

exports[`baseFilterToFilterDto > should return correctly translated simple filter 1`] = `
{
  "disabled": false,
  "isCascading": false,
  "jaql": {
    "column": "Gender",
    "datatype": "text",
    "dim": "[Commerce.Gender]",
    "filter": {
      "members": [
        "Male",
      ],
      "multiSelection": true,
    },
    "table": "Commerce",
    "title": "Gender",
  },
}
`;
